### PR TITLE
chore(types): make Props param in ShorthandValue and ShorthandCollection required

### PR DESCRIPTION
**Pull request by layershifter**
# BREAKING CHANGES

Only if you have used `ShorthandValue` or `ShorthandCollection` types.

### Before
```tsx
const item: ShorthandValue = {}
const items: ShorthandCollection = []
```

### After
```tsx
const item: ShorthandValue<ListItemProps> = {}
const items: ShorthandCollection<ListItemProps> = []
```

***

This PR makers `P` in `ShorthandValue<P>` required, this will allow to get better autocomplete, improve types and generate better prop definitions.

### CODE REVIEW

File: packages/react/src/components/Menu/Menu.tsx [hunk 1]
```diff
@@ -15,14 +15,15 @@ import {
...
-import MenuItem from './MenuItem'
+import MenuItem, { MenuItemProps } from './MenuItem'
 import { menuBehavior } from '../../lib/accessibility'
 import { Accessibility } from '../../lib/accessibility/types'
 import { ReactAccessibilityBehavior } from '../../lib/accessibility/reactTypes'
 
 import { ComponentVariablesObject, ComponentSlotStylesPrepared } from '../../themes/types'
 import { WithAsProp, ShorthandCollection, ShorthandValue, withSafeTypeForAs } from '../../types'
 import MenuDivider from './MenuDivider'
+import { IconProps } from '@stardust-ui/react'
```
Comment (miroslavstastny) [inline packages/react/src/components/Menu/Menu.tsx]: wrong import (+3 other places)
image

File: packages/react/src/components/Slider/Slider.tsx [hunk 2]
```diff
@@ -83,7 +83,7 @@ export interface SliderProps
   getA11yValueMessageOnChange?: (props: SliderProps) => string
 
   /** Shorthand for the input component. */
-  input?: ShorthandValue
+  input?: ShorthandValue<HtmlInputProps>
```
Comment (miroslavstastny) [inline packages/react/src/components/Slider/Slider.tsx]: 😮

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 3]
```diff
@@ -46,10 +46,10 @@ export interface TreeItemProps extends UIComponentProps, ChildrenComponentProps
   defaultOpen?: boolean
 
   /** The index of the item among its sibbling */
-  index: number
+  index?: number
 
   /** Array of props for sub tree. */
-  items?: ShorthandValue[]
+  items?: ShorthandValue<TreeItemProps>[]
```
Comment (miroslavstastny) [inline packages/react/src/components/Tree/TreeItem.tsx]: What's the reason for using `ShorthandValue[]`?
image

File: packages/react/src/components/Menu/Menu.tsx (hunk 1 above)
Comment (layershifter) [reply packages/react/src/components/Menu/Menu.tsx]: Great catch 👍

File: packages/react/src/components/Tree/TreeItem.tsx (hunk 3 above)
Comment (layershifter) [reply packages/react/src/components/Tree/TreeItem.tsx]: No reasons, updated it 👍

File: packages/react/src/components/Slider/Slider.tsx (hunk 2 above)
Comment (layershifter) [reply packages/react/src/components/Slider/Slider.tsx]: Oops, should be `BoxProps` actually

File: docs/src/examples/components/Chat/Content/ChatExampleReactionGroupMeReacting.shorthand.tsx [hunk 4]
```diff
@@ -1,12 +1,12 @@
 import * as React from 'react'
-import { Avatar, Chat } from '@stardust-ui/react'
+import { Avatar, Chat, ChatItemProps, ShorthandCollection } from '@stardust-ui/react'
 
 const reactions = [
```
Comment (mnajdova) [inline docs/src/examples/components/Chat/Content/ChatExampleReactionGroupMeReacting.shorthand.tsx]: Let's type the reactions as well. Wondering why this wasn't catch... :(

File: packages/react/src/components/Chat/ChatItem.tsx [hunk 5]
```diff
@@ -35,13 +35,13 @@ export interface ChatItemProps extends UIComponentProps, ChildrenComponentProps
...
   /** Chat items can have a gutter. */
-  gutter?: ShorthandValue
+  gutter?: ShorthandValue<BoxProps>
 
   /** Indicates whether the content is positioned at the start or the end. */
   contentPosition?: 'start' | 'end'
 
   /** Chat items can have a message. */
-  message?: ShorthandValue
+  message?: ShorthandValue<ChatMessageProps>
```
Comment (mnajdova) [inline packages/react/src/components/Chat/ChatItem.tsx]: Actually this should be `BoxProps`, see line 109. We can easily do some miss match here.. :(

File: docs/src/examples/components/Chat/Content/ChatExampleReactionGroupMeReacting.shorthand.tsx (hunk 4 above)
Comment (layershifter) [reply docs/src/examples/components/Chat/Content/ChatExampleReactionGroupMeReacting.shorthand.tsx]: Done 👍

File: packages/react/src/components/Chat/ChatItem.tsx (hunk 5 above)
Comment (layershifter) [reply packages/react/src/components/Chat/ChatItem.tsx]: Great catch 👍 

I am bit disappointed that we don't have strict checks there, but at least we will be able to improve them and catch such cases

File: docs/src/components/ComponentBestPractices.tsx [hunk 6]
```diff
@@ -1,16 +1,16 @@
...
-import _ from 'lodash'
-import { Box, Flex, Header, Icon, List, Segment, ShorthandValue } from '@stardust-ui/react'
+import * as _ from 'lodash'
+import { Box, Flex, Header, Icon, List, Segment, ShorthandCollection } from '@stardust-ui/react'
 import { Extendable } from 'src/types'
 
 export type ComponentBestPracticesProps = Extendable<{
-  doList?: ShorthandValue<{}>[]
-  dontList?: ShorthandValue<{}>[]
+  doList?: ShorthandCollection<{}>
```
Comment (mnajdova) [inline docs/src/components/ComponentBestPractices.tsx]: Do we really need the empty object here?

File: packages/react/src/components/Dropdown/Dropdown.tsx [hunk 7]
```diff
@@ -115,21 +117,21 @@ export interface DropdownProps
   inline?: boolean
 
   /** Array of props for generating list options (Dropdown.Item[]) and selected item labels(Dropdown.SelectedItem[]), if it's a multiple selection. */
-  items?: ShorthandCollection
+  items?: ShorthandCollection<DropdownItemProps>
```
Comment (mnajdova) [inline packages/react/src/components/Dropdown/Dropdown.tsx:120]: This is also use in the `DropdownSelectedItem.create` factory. How should we handle this one?

File: packages/react/src/components/Dropdown/DropdownItem.tsx [hunk 8]
```diff
@@ -24,13 +24,13 @@ export interface DropdownItemProps extends UIComponentProps<DropdownItemProps> {
   accessibilityItemProps?: any
 
   /** Item's content. */
-  content?: ShorthandValue
+  content?: ShorthandValue<BoxProps>
 
   /** Item's header. */
-  header?: ShorthandValue
+  header?: ShorthandValue<ImageProps>
```
Comment (mnajdova) [inline packages/react/src/components/Dropdown/DropdownItem.tsx]: Oops you switched the Image and box props :)

File: packages/react/src/components/Dropdown/DropdownSelectedItem.tsx [hunk 9]
```diff
@@ -30,13 +31,13 @@ export interface DropdownSelectedItemProps extends UIComponentProps<DropdownSele
   active?: boolean
 
   /** Header of the selected item. */
-  header?: ShorthandValue
+  header?: ShorthandValue<HeaderProps>
```
Comment (mnajdova) [inline packages/react/src/components/Dropdown/DropdownSelectedItem.tsx]: This should be `BoxProps` too

File: packages/react/src/components/Reaction/Reaction.tsx [hunk 10]
```diff
@@ -29,17 +29,17 @@ export interface ReactionSlotClassNames {
...
    */
   accessibility?: Accessibility
 
   /** A reaction can have icon for the indicator of the reaction. */
-  icon?: ShorthandValue
+  icon?: ShorthandValue<IconProps>
 
   /** A reaction can have content shown next to the icon. */
-  content?: ShorthandValue
+  content?: ShorthandValue<BoxProps>
```
Comment (mnajdova) [inline packages/react/src/components/Reaction/Reaction.tsx]: I think we can actually remove this, as we already have it defined when we are extending `ContentComponentProps<ShorthandValue<BoxProps>>`

File: packages/react/src/components/Tree/TreeItem.tsx [hunk 11]
```diff
@@ -43,10 +44,10 @@ export interface TreeItemProps extends UIComponentProps, ChildrenComponentProps
   exclusive?: boolean
 
   /** The index of the item among its sibbling */
-  index: number
+  index?: number
 
   /** Array of props for sub tree. */
-  items?: ShorthandValue[]
+  items?: ShorthandCollection<TreeItemProps>
```
Comment (mnajdova) [inline packages/react/src/components/Tree/TreeItem.tsx]: This should be `TreeProps`

Comment (mnajdova) [reply packages/react/src/components/Tree/TreeItem.tsx]: But then it won't be Collection.. Not sure if it should be changed, but it seems confusing as it is used in the `Tree.create(items, ..)` factory

File: docs/src/components/ComponentBestPractices.tsx (hunk 6 above)
Comment (layershifter) [reply docs/src/components/ComponentBestPractices.tsx]: It should be `ShorthandCollection<ListItemProps>`, fixed 👍

File: packages/react/src/components/Dropdown/DropdownItem.tsx (hunk 8 above)
Comment (layershifter) [reply packages/react/src/components/Dropdown/DropdownItem.tsx]: Fixed ❤️

File: packages/react/src/components/Dropdown/Dropdown.tsx (hunk 7 above)
Comment (layershifter) [reply packages/react/src/components/Dropdown/Dropdown.tsx:120]: I don't have a good idea, left a comment as we discussed offline

File: packages/react/src/components/Dropdown/DropdownSelectedItem.tsx (hunk 9 above)
Comment (layershifter) [reply packages/react/src/components/Dropdown/DropdownSelectedItem.tsx]: Fixed 👍

File: packages/react/src/components/Tree/TreeItem.tsx (hunk 11 above)
Comment (layershifter) [reply packages/react/src/components/Tree/TreeItem.tsx]: As `Tree` has `mappedArrayProp` it should be:
```suggestion
  items?: ShorthandValue<TreeProps> | ShorthandCollection<TreeItemProps>
```

File: packages/react/src/components/Chat/ChatMessage.tsx [hunk 12]
```diff
@@ -83,7 +84,7 @@ export interface ChatMessageProps
   onFocus?: ComponentEventHandler<ChatMessageProps>
 
   /** Reaction group applied to the message. */
-  reactionGroup?: ShorthandValue
+  reactionGroup?: ShorthandValue<ReactionGroupProps>
```
Comment (mnajdova) [inline packages/react/src/components/Chat/ChatMessage.tsx]: Shouldn't this be: `ShorthandValue<ReactionGroupProps> | ShorthandCollection<ReactionProps>`

**Review by mnajdova — APPROVED** [review]
Don't forget to add changelog with **BREAKING CHANGES** entry.

**Pull request merged**